keps/sig-storage/20190120-ExecutionHook-apichange.md

```diff
@@ -0,0 +1,723 @@
+* When PostAction has been run successfully on all selected containers in the pods, a PostActionSucceed summary status will be set to true. If PostAction fails on any container, PostActionSucceed will be set to false but the hook controller will continue to run the command on the remaining containers. When PostActionSucceed summary status is set to true, the hook controller will finally delete the ExecutionHook.
+* There could be multiple pods and containers selected for one hook. The hook controller needs to wait for hook commands to be done on all selected containers. Only if command runs successfully on all containers on all pods, the overall status is successful.
+* If there are more than 1 ExecutionHooks, the snapshot controller will have to delete all hooks after cutting the snapshot.
+* PreAction and PostAction hooks are not required. While most applications require both PreAction and PostAction hooks, some applications such as Mysql only require a PreAction hook. When a PreAction command is run in a mysql session to freeze the application, the PostAction command must be run from the same session to unfreeze it. If the session is closed after the PreAction command, the mysql application will automatically be unfrozen. Therefore a PreAction hook for mysql needs to be running in the background after running the freeze command, and PostAction command will be run within the same session after the snapshot is cut (or closing the session to unfreeze it). For most other applications, PreAction and PostAction do not have to happen in the same session so we will have two separate handlers.
+* The ExecutionHook controller will guarantee that a PreAction or PostAction command is issued, however, it cannot guarantee the command will succeed because that is determined by the application.
+* It is recommended that the user adds `PostAction` command inside the `PreAction` script and makes sure that `PostAction` is automatically run after the `PreAction` command has completed successfully for a period of time, same as the ExpirationTimeout.
+* Both PreAction and PostAction should be idempotent.
+
+Here is an example of an ExecutionHookTemplate:
+```
+apiVersion: storage.k8s.io/v1alpha1
+kind: ExecutionHookTemplate
+metadata:
+  name: hook-template-demo
+policy: ExecuteOnce
+preAction:
+  exec:
+    command: [“run_quiesce.sh”]
+  actionTimeoutSeconds: 10
+postAction:
+  exec:
+    command: [“run_unquiesce.sh”]
+  actionTimeoutSeconds: 10
+expirationTimeSeconds: 30
+```
+
+Here is an example of how the ExecutionHookInfos is used in a VolumeSnapshot object:
+```
+apiVersion: snapshot.storage.k8s.io/v1alpha1
+kind: VolumeSnapshot
+metadata:
+  name: snapshot-demo
+spec:
+  snapshotClassName: csi-snapshot-class
+  source:
+    name: pvc-demo
+    kind: PersistentVolumeClaim
+  executionHookInfos:
+    - executionHookTemplateName: hook-template-demo
```

Comment:
should have pod/container information too?

Reply:
Sure, I'll add that.  If pod/container info is not set, the snapshot controller can automatically find them too.

Reply:
I'll change to this:
```
  executionHookInfos:
    - executionHookTemplateName: new-hook-template
      podContainerNamesList:
        - podName: mysql
          containerNames: ["mysql"]
```